initrd/etc/luks-functions

```diff
@@ -267,24 +268,37 @@ prepare_thumb_drive()
 
 select_luks_container()
 {
+	#TODO: extend logic to prompt for block devices with model if multiple LUKS are found on block device instead of partitions
+	# Then feed luks with those partitions so that reencrypt and passphrase change can use passphrase to test all selected
 	TRACE_FUNC
 	if [ -s /boot/kexec_key_devices.txt ]; then
 		DEBUG "Reusing known good LUKS container device from /boot/kexec_key_devices.txt"
-		DEBUG "LUKS container device: $(cut -d ' ' -f1 /boot/kexec_key_devices.txt)"
-		LUKS=$(cut -d ' ' -f1 /boot/kexec_key_devices.txt)
-	else
+		LUKS=$(cut -d ' ' -f1 /boot/kexec_key_devices.txt) 
+		DEBUG "LUKS container device: $(echo $LUKS)"
+		# LUKS variable not exported yet, prompt for LUKS device
+	elif [ -z "$LUKS" ]; then
 		list_luks_devices > /tmp/luks_devices.txt
 		#if /tmp/luks_devices.txt exists and is not empty
 		if [ -s /tmp/luks_devices.txt ]; then
 			file_selector "/tmp/luks_devices.txt" "Select LUKS container device"
 			if [ "$FILE" == "" ]; then
 				return 1
 			else
-				LUKS=$FILE
-				detect_boot_device
-				mount -o remount,rw /boot
-				echo "$LUKS $(cryptsetup luksUUID $LUKS)" >/boot/kexec_key_devices.txt
-				mount -o remount,ro /boot
+				#TODO: What about BRTFS multi LUKS setup of QubesOS?
```

Comment:
I think you mean LVM, not BTRFS.  IIRC Qubes uses multiple LVM volumes, including for the VMs.  BTRFS supports multiple datasets but LUKS encryption is done at the block device layer.

Do we really need to interact with LUKS volumes other than the Qubes root filesystem?  Admittedly I have not looked into how the VMs' filesystems are encrypted, but I would have guessed that they used separate keys held on the root volume.  I wouldn't have expected it to use the same key as the root, because then Qubes would need to keep that secret in memory to unlock VM roots later.  But I haven't looked into that myself yet.

If we do need to do this, then I suppose we need to figure out what will happen if you create a new qube later, since Qubes won't know to do this to its partition(s).

Reply:
@JonathonHall-Purism no, I really meant BTRFS (typo) and multi LUKS (rootfs and SWAP are in seperate LUKS containers), so that TPM DUK needs to know both LUKS and be able to unlock both (in same block device, /dev/sda3 /dev/sda5) to be set, same for reencryption.

There is confusion here.
STD Q4.2 deploys multiple lvms on vg on single LUKS partition, worked before and now. What changes here is able to deal with multiple LUKS all the way from reencryption, setting new passphrase and TPM DUK, not taking for granted LUKS key slots where Recovery Key is stored, dealing with legacy prior Heads hardoding DUK in LUKS key slot 1 and taking for granted Disk Recovery Key was in slot 0.

Reply:
OK, so the goal is to support a setup with one LUKS container for rootfs (possibly LVM, doesn't matter), and a second LUKS container for swap.  That makes sense now.

I'm not sure whether it's the default for Qubes 4.2 or a custom setup, but I can see a reasoning that if you can set up two LUKS partitions with the same passphrase and the default disk unlock would unlock both of them automatically (from one passphrase entry), then TPM DUK should be able to do it too.

So I think I get the motivation here.